### ISSUE TITLE
Support for naming sub components in Model for Swagger

### ISSUE THREAD

**Issue by cate-apt** [http4k/http4k#661]
Using Contracts Library to generate a Swagger for my application, I used definitionId to add name for the Model, however when the model have a subcomponent that is a data class, the name of the class was replaces with "object+number". Something like this:

    Request:
      type: object
      properties:
        test:
          type: string
          example: test1
        test2:
          type: string
          example: test2
        test3:
          type: string
          example: test3
        test4:
          $ref: '#/components/schemas/object-1213812167'
    
    object-1213812167:
      type: object
      properties:
        test5:
          $ref: '#/components/schemas/object1084711580'
        test6:
          type: array
          items:
            $ref: '#/components/schemas/object-1127775706'
        test7:
          type: array
          items:
            $ref: '#/components/schemas/object887621248'
        test8:
          type: string
          example: test8

The contract have this in the receiving:
        receiving(RequestMeta(Request(POST, "").with(requestLens of Request()), "Request"))

And the Data Class for Request is something like this:

data class OnboardingRequest(
    val test1: String,
    val test2: String,
    val test3: String,
    val test4: ParamaterTest
)

data class ParamaterTest(
    val test5: ParamTest5,
    val test6: List<ParamTest6>,
    val test7: List<ParamTest7,
    val test8: String
)

I used the solution for naming model as suggested in Issue #88 for the high level models. How can the sub component in the model can have proper naming?

**Comment by daviddenton** [comment]
Sorry - unfortunately this isn't possible with OpenApi2 or 3.

However, if you using OpenApi3 correctly with Jackson, it's much easier because the spec gets created with names that match the class name.

**Comment by cate-apt** [comment]
I tried to add Jackson in OpenAPI3 declaration, I used it like this:

    val openApi3 = OpenApi3(ApiInfo("Project", "1.0", "Application"),
        Jackson,
        errorResponseRenderer = BadRequestErrorResponseRenderer()
    )

This one doesn't use the class as object name, I need to add definitionId in the receiving and returning in the Contract. I even tried using the ConfigurableJackson with this configuration:

object ConfigJackson : ConfigurableJackson(
    KotlinModule()
        .asConfigurable()
        .withStandardMappings()
        .done()
        .deactivateDefaultTyping()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        .configure(DeserializationFeature.FAIL_ON_IGNORED_PROPERTIES, false)
        .configure(DeserializationFeature.ACCEPT_EMPTY_ARRAY_AS_NULL_OBJECT, false)
        .configure(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS, true)
        .configure(DeserializationFeature.USE_BIG_INTEGER_FOR_INTS, true)
        .configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false)
        .setSerializationInclusion(JsonInclude.Include.NON_NULL)
        .setSerializationInclusion(JsonInclude.Include.NON_EMPTY)
)

Still have the same issue. Do I miss something so the object will use the data class names?

**Comment by daviddenton** [comment]
Strange - I can't see anything obviously wrong. It might be worth looking at this test to see where your differences lie, because it does exactly what you're trying to do:

https://github.com/http4k/http4k/blob/d56978c1d05ca9a28b90e617b9e5163a7f166284/http4k-contract/src/test/kotlin/org/http4k/contract/openapi/v3/OpenApi3AutoTest.kt

... specifically this line which sets the name: https://github.com/http4k/http4k/blob/d874f0c0e9e06ad7a2b7adea18666f74af72540b/http4k-contract/src/test/kotlin/org/http4k/contract/ContractRendererContract.kt#L194

... and generates:
```
      "someOtherId": {
        "properties": {
          "string": {
            "example": "s",
            "description": null,
            "type": "string"
          },
          "child": {
            "$ref": "#/components/schemas/ArbObject1",
            "example": null,
            "description": null
          },
          "numbers": {
            "items": {
              "type": "number"
            },
            "example": [
              1
            ],
            "description": null,
            "type": "array"
          },
          "bool": {
            "example": true,
            "description": null,
            "type": "boolean"
          }
        },
        "example": {
          "string": "s",
          "child": {
            "anotherString": "bar"
          },
          "numbers": [
            1
          ],
          "bool": true
        },
        "description": null,
        "type": "object",
        "required": [
          "bool",
          "numbers",
          "string"
        ]
      },

```

**Comment by cate-apt** [comment]
Using definitionId works for the high-level model like what you show the someOtherId, what doesn't work for me is the ArbObject1 it was objectName+Number, tried doing the same structure with this link:

http4k/http4k-contract/src/test/kotlin/org/http4k/contract/ContractRendererContract.kt

I tried adding the codes from the link above in my project and the result is null:
    someOtherId:
      type: object
      properties: {}

Need to use RequestMeta to be able to get a value:

        receiving(RequestMeta(Request(POST, "")
            .with(Body.auto<ArbObject2>().toLens() of ArbObject2(
                "s",
                ArbObject1(Foo.bar),
                listOf(1),
                true
            )),"someOtherId"))

enum class Foo {
    bar, bing
}


data class ArbObject1(val anotherString: Foo)
data class ArbObject2(val string: String, val child: ArbObject1?, val numbers: List<Int>, val bool: Boolean)

The result also have objectName and number for the sub component
    someOtherId:
      type: object
      properties:
        string:
          type: string
          example: s
        child:
          $ref: '#/components/schemas/object-1479447290'
        numbers:
          type: array
          items:
            type: integer
            example: 1
        bool:
          type: boolean
          example: true

This is the same code as what in the link, but the result is different.

**Comment by cate-apt** [comment]
ping .-)

**Comment by daviddenton** [comment]
Something that might help here is the workaround to #661 - where you can set a custom SchemaModelNamer. https://github.com/http4k/http4k/issues/681

**Comment by daviddenton** [comment]
I think this will be fixed (or at least workaroundable in 4.21.1.0) as we now have the ability to provide a custom prefix for models in a particular example, which unlocks the ability to "name" submodels in a particular way in concert with [this](https://github.com/http4k/http4k/issues/661#issuecomment-1058313944) hint. 

This ia all we can do really, so closing this off as we can workaround,
